### ISSUE TITLE
LaTeX spacing around brackets is off

### ISSUE THREAD

**Issue by TomFryers** [sympy/sympy#23286]
```python
>>> from sympy import *
>>> print(printing.latex(sin(3)))
\sin{\left(3 \right)}
```
If we now put this into LaTeX:
```latex
\documentclass{article}
\begin{document}
\[\sin(3)\] % comparison
\[\sin{\left(3 \right)}\]
\end{document}
```
we get [this output](https://i.imgur.com/rOgpiIP.png). There is too much space between ‘sin’ and the brackets. [This is caused by `\left` and `\right`.](https://tex.stackexchange.com/a/173740/200616) There are some possible solutions discussed [here](https://tex.stackexchange.com/questions/2607/spacing-around-left-and-right/2610#2610). Without extra packages, the most feasible is probably to replace all `\left(`s with `\mathopen{}\left(`s and all `\right)`s with `\right)\mathclose{}`s (and the same with other bracket types). This makes the LaTeX source less human-readable, but it may be possible to mitigate this it by not using automatic bracket sizing in some cases where it is known to be unnecessary.

**Comment by KuldeepBorkar** [comment]
Hi, @TomFryers 
I would like to work on this issue.

There are few things I noticed;

--> Here according to your answer instead of
`\sin{\left(3 \right)}`
Below code will have no extra space right? 
`\sin{\mathopen{}\left(3 \right)\mathclose{}}`

but in my case this worked
`\sin\mathopen{}{\left(3 \right)}\mathclose{}` 
(i.e. Replacing `{\left(` with `\mathopen{}{\left(` and `\right)}` with `\right)}\mathclose{}`)

--> But, the major thing you are suggesting is to enclose all the `\left`'s and `\right`'s in `mathopen{}` and `mathclose{}` right?

I guess I should probably make changes in `sympy\printing\latex.py`
and replace all the statements where `\left` and `\right` is used, to be enclosed under `mathopen{}` and `mathclose{}` so that unnecessary spaces can be avoided.
If someone from the `SymPy` Community could confirm this once I would like to implement the changes : )

**Comment by TomFryers** [comment]
Yes, you are right, just replacing as I said doesn’t quite work because the extra curly brackets around them reset the spacing. Moving them inwards as you did should work.

**Comment by asmeurer** [comment]
Do \left and \right have this problem when they actually are needed?

It would be nice to only produce \left and \right when they are needed but that seems to be a difficult problem. I guess we could use some basic heuristics and fallback when necessary. 

Also, does MathJax suffer from this problem? I've never noticed it before.

**Comment by TomFryers** [comment]
Yes, they always have poor spacing. MathJax seems only to have the problem when the extra curly brackets are present, but for KaTeX and LaTeX, it’s an issue with or without them.

**Comment by gschintgen** [comment]
I didn't know this either (limited experience via pythontex), but I'd much rather see some special heuristics to avoid left/right or a note in the docs about workarounds (e.g. using the package mleftright) than having mathopen/mathclose all over the place. In some cases the additional space may well be beneficial: https://tex.stackexchange.com/a/440016

**Comment by asmeurer** [comment]
I expect if we do this it should only be for things that are known to look worse without it. It seems like that might just be "functions" and not operators like \sum, but maybe there are other exceptions as well. 

I'm not convinced yet if this should be the default, but it sounds reasonable to at least have an option for it. A first step would be to refactor the code in latex.py so that places that insert `\left(...\right)` do so via a `_parenthesize` function (c.f. `parenthesize` in the pretty printer). This function could then be passed an option to enable adding `\mathclose{}`, which could itself be overridden with a flag to latex() (and we can debate what the default for that flag would be). 

This refactory would also open the possibility for heuristics to avoid \left and \right. That would be generally useful, because it would be nice to make the output more human readable. But it sounds like even if we had perfect heuristics, that wouldn't really solve this problem because the spacing is wrong even when \left and \right are required. 

>  MathJax seems only to have the problem when the extra curly brackets are present

Can you clarify what you mean here? Are you saying that `\sin{\left(x\right)}` is different from `\sin\left(x\right)`? Or something else? Also some screenshots showing the differences would be useful.

**Comment by TomFryers** [comment]
> Are you saying that \sin{\left(x\right)} is different from \sin\left(x\right)?

Yes. It seems that the curly brackets quite often change the output.

I’ve got some screenshots of different cases. The source code is
```latex
\documentclass{article}
\begin{document}
Sine function:\\
\verb|\sin(y)i|
\[\sin(y)i\]
\verb|\sin\left(y \right)i|
\[\sin\left(y \right)i\]
\verb|\sin{\left(y \right)}i|
\[\sin{\left(y \right)}i\]
\verb|\sin\mathopen{}\left(y \right)\mathclose{}i|
\[\sin\mathopen{}\left(y \right)\mathclose{}i\]
\verb|\sin\mathopen{}{\left(y \right)}\mathclose{}i|
\[\sin\mathopen{}{\left(y \right)}\mathclose{}i\]
Arbitrary function (or product):\\
\verb|f(y)i|
\[f(y)i\]
\verb|f\left(y \right)i|
\[f\left(y \right)i\]
\verb|f{\left(y \right)}i|
\[f{\left(y \right)}i\]
\verb|f\mathopen{}\left(y \right)\mathclose{}i|
\[f\mathopen{}\left(y \right)\mathclose{}i\]
\verb|f\mathopen{}{\left(y \right)}\mathclose{}i|
\[f\mathopen{}{\left(y \right)}\mathclose{}i\]
Integral:\\
\verb|\int (x + 1)\, dx|
\[\int (x + 1)\, dx\]
\verb|\int \left(x + 1 \right)\, dx|
\[\int \left(x + 1 \right)\, dx\]
\verb|\int {\left(x + 1 \right)}\, dx|
\[\int {\left(x + 1 \right)}\, dx\]
\verb|\int \mathopen{}\left(x + 1 \right)\mathclose{}\, dx|
\[\int \mathopen{}\left(x + 1 \right)\mathclose{}\, dx\]
\verb|\int \mathopen{}{\left(x + 1 \right)}\mathclose{}\, dx|
\[\int \mathopen{}{\left(x + 1 \right)}\mathclose{}\, dx\]
\end{document}
```

PDFLaTeX and XeLaTeX both generate [this](https://i.imgur.com/6AHkEhD.png).

Currently, SymPy doesn’t add curly brackets to the integral, but does to the functions; it outputs the third, third and second options.

[KaTeX’s output](https://i.imgur.com/vl8Rz3T.png) is essentially the same, as far as I can tell. I’ll try MathJax later.

**Comment by asmeurer** [comment]
Can you upload the images directly to the GitHub comment rather than using imgur? That will make it easier to view and compare them.

**Comment by TomFryers** [comment]
> Can you upload the images directly to the GitHub comment rather than using imgur? That will make it easier to view and compare them.

I’ve done that now (edited previous posts).

**Comment by TomFryers** [comment]
OK, here it is with MathJax:
![MathJax](https://i.imgur.com/gvZthgJ.png)

The only difference seems to be that MathJax has less space than everything else after `\right)` with no curly brackets.

**Comment by asmeurer** [comment]
It looks like the extra space with curly brackets is there for `f` but not for `\sin`. I'm not entirely sure how curly brackets work in LaTeX, but I guess the fact that `f` is not a command means that they end up being interpreted differently. It looks like `latex(f(x))` does produce those curly brackets, but it sounds like we should modify it to only use them after a command. It would be nice to confirm that this is indeed what is going on, though. 

As far as aesthetics, I personally don't think the extra space looks good in any of the examples. Even the integral looks better without it (although it's extremely subtle). Maybe a sum would be different like in the StackExchange question.
